File: CS Template/MetaWearBoard.cs
using static MbientLab.MetaWear.Functions;

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Windows.Devices.Bluetooth.GenericAttributeProfile;
using System.Runtime.InteropServices;
using Windows.Devices.Bluetooth;
using System.Runtime.InteropServices.WindowsRuntime;
using MbientLab.MetaWear.Core;

namespace MbientLab.MetaWear.Template {
    /// <summary>
    /// Wrapper class that Sets up the C++ API to be used in C# following the steps outlined in the C++ documentation:
    /// https://mbientlab.com/cppdocs/latest/btlecommunication.html
    /// </summary>
    public sealed class MetaWearBoard {
        private static Dictionary<ulong, MetaWearBoard> instances= new Dictionary<ulong, MetaWearBoard>();
        public static MetaWearBoard getMetaWearBoardInstance(BluetoothLEDevice btleDevice) {
            MetaWearBoard board;

            if (!instances.TryGetValue(btleDevice.BluetoothAddress, out board)) {
                board = new MetaWearBoard(btleDevice);
                instances.Add(btleDevice.BluetoothAddress, board);
            }
            return board;
        }

        private static Dictionary<IntPtr, MetaWearBoard> cppApiBoards = new Dictionary<IntPtr, MetaWearBoard>();
        /// <summary>
        /// Writes a value to a GATT characteristic
        /// </summary>
        /// <param name="caller">Object calling this function</param>
        /// <param name="gattCharPtr">Pointer to a <see cref="MbientLab.MetaWear.Core.GattCharacteristic"/></param>
        /// <param name="value">Pointer to a byte array containing the value</param>
        /// <param name="length">Number of bytes</param>
        private async static void writeCharacteristic(IntPtr caller, IntPtr gattCharPtr, IntPtr value, byte length) {
            byte[] managedArray = new byte[length];
            Marshal.Copy(value, managedArray, 0, length);
            
            MetaWearBoard board;
            if (cppApiBoards.TryGetValue(caller, out board)) {
                try {
                    var charAsGuid = Marshal.PtrToStructure<MbientLab.MetaWear.Core.GattCharacteristic>(gattCharPtr).toGattCharGuid();
                    var status = await board.btleDevice.GetGattService(charAsGuid.serviceGuid).GetCharacteristics(charAsGuid.guid).FirstOrDefault()
                        .WriteValueAsync(managedArray.AsBuffer(), GattWriteOption.WriteWithoutResponse);

                    System.Diagnostics.Debug.WriteLineIf(status != GattCommunicationStatus.Success, "Cannot write gatt characteristic " + charAsGuid.ToString());
                } catch (Exception e) {
                    System.Diagnostics.Debug.WriteLine("Error writing gatt characteristic: " + e.Message);
                }
            }
        }
        /// <summary>
        /// Reads the value from a GATT characteristic
        /// </summary>
        /// <param name="caller">Object calling this function</param>
        /// <param name="gattCharPtr">Pointer to a <see cref="MbientLab.MetaWear.Core.GattCharacteristic"/></param>
        private async static void readCharacteristic(IntPtr caller, IntPtr gattCharPtr) {
            MetaWearBoard board;

            if (cppApiBoards.TryGetValue(caller, out board)) {
                try {
                    var charAsGuid = Marshal.PtrToStructure<MbientLab.MetaWear.Core.GattCharacteristic>(gattCharPtr).toGattCharGuid();
                    var result = await board.btleDevice.GetGattService(charAsGuid.serviceGuid).GetCharacteristics(charAsGuid.guid).FirstOrDefault()
                        .ReadValueAsync();

                    if (result.Status == GattCommunicationStatus.Success) {
                        mbl_mw_connection_char_read(caller, gattCharPtr, result.Value.ToArray(), (byte)result.Value.Length);
                    } else {
                        System.Diagnostics.Debug.WriteLine("Cannot read gatt characteristic " + charAsGuid.ToString());
                    }
                } catch (Exception e) {
                    System.Diagnostics.Debug.WriteLine("Error reading gatt characteristic: " + e.Message);
                }
            }
        }

        private Windows.Devices.Bluetooth.GenericAttributeProfile.GattCharacteristic notifyChar= null;
        private BluetoothLEDevice btleDevice;
        /// <summary>
        /// Pointer representing the MblMwMetaWearBoard struct created by the C++ API
        /// </summary>
        public IntPtr cppBoard { get; }
        /// <summary>
        /// C# wrapper around the MblMwBtleConnection struct 
        /// </summary>
        private BtleConnection btleConn;
        
        private MetaWearBoard(BluetoothLEDevice btleDevice) {
            this.btleDevice = btleDevice;

            btleConn = new BtleConnection();
            btleConn.writeGattChar = new Fn_IntPtr_IntPtr_ByteArray(writeCharacteristic);
            btleConn.readGattChar = new Fn_IntPtr_IntPtr(readCharacteristic);

            cppBoard = mbl_mw_metawearboard_create(ref btleConn);
            cppApiBoards.Add(cppBoard, this);
        }

        /// <summary>
        /// Initialize the API
        /// </summary>
        /// <returns>Status value from calling <see cref="mbl_mw_metawearboard_initialize(IntPtr, Fn_IntPtr_Int)"/></returns>
        public async Task<int> Initialize() {
            if (notifyChar == null) {
                notifyChar = btleDevice.GetGattService(GattCharGuid.METAWEAR_NOTIFY_CHAR.serviceGuid).GetCharacteristics(GattCharGuid.METAWEAR_NOTIFY_CHAR.guid).FirstOrDefault();
                notifyChar.ValueChanged += notifyHandler;
                await notifyChar.WriteClientCharacteristicConfigurationDescriptorAsync(GattClientCharacteristicConfigurationDescriptorValue.Notify);
            }

            
            return await Task.Run(() => {
                int initStatus = -1;

                Fn_IntPtr_Int initDelegate = new Fn_IntPtr_Int((caller, status) => {
                    initStatus = status;
                });
                mbl_mw_metawearboard_initialize(cppBoard, initDelegate);

                return initStatus;
            });
        }

        /// <summary>
        /// Free the memory allocated by the C++ API
        /// </summary>
        public void Free() {
            mbl_mw_metawearboard_free(cppBoard);

            notifyChar.ValueChanged -= notifyHandler;
            notifyChar = null;

            instances.Remove(btleDevice.BluetoothAddress);
            cppApiBoards.Remove(cppBoard);
        }

        /// <summary>
        /// Handles notifications from the MetaWear notify characteristic
        /// </summary>
        private void notifyHandler(Windows.Devices.Bluetooth.GenericAttributeProfile.GattCharacteristic gattCharChanged, GattValueChangedEventArgs obj) {
            byte[] response = obj.CharacteristicValue.ToArray();
            mbl_mw_connection_notify_char_changed(cppBoard, response, (byte)response.Length);
        }
    }
}

File: Cpp Template/MetaWearBoard.cpp
#include "pch.h"

#include "MetaWearBoard.h"
#include "metawear/platform/btle_connection.h"

#include <unordered_map>

using namespace Platform;
using namespace Cpp_Template;
using namespace Windows::Devices::Bluetooth;
using namespace Windows::Devices::Bluetooth::GenericAttributeProfile;
using namespace Windows::Security::Cryptography;

using std::unordered_map;

static unordered_map<uint64, MetaWearBoard^> instances;
static unordered_map<const void*, MetaWearBoard^> cppToWrapper;
MblMwBtleConnection MetaWearBoard::conn = { MetaWearBoard::writeGattCharacteristic, MetaWearBoard::readGattCharacteristic };

static inline Guid byteArrayToGuid(const uint8_t* uuids) {
    Array<byte>^ lower = ref new Array<byte>(8);
    for (int i = 0; i < lower->Length; i++) {
        lower[i] = *(uuids + 128 + i);
    }

    return Guid(*((uint32*) (uuids)), *((uint16*) uuids + 2), *((uint16*)uuids + 3), lower);
}

void MetaWearBoard::writeGattCharacteristic(const void* caller, const MblMwGattChar* characteristic, const uint8_t* value, uint8_t length) {
    Array<byte>^ wrapper = ref new Array<byte>(length);
    for (int i = 0; i < length; i++) {
        wrapper[i] = value[i];
    }

    Guid service = byteArrayToGuid((const uint8_t*)characteristic), guidChar = byteArrayToGuid((const uint8_t*)(&characteristic->uuid_high));
    auto dest = cppToWrapper.at(caller)->device->GetGattService(service)->GetCharacteristics(guidChar)->First()->Current;
    concurrency::create_task(dest->WriteValueAsync(CryptographicBuffer::CreateFromByteArray(wrapper), GattWriteOption::WriteWithoutResponse)).then([](GattCommunicationStatus status) {
        if (status != GattCommunicationStatus::Success) {
            OutputDebugString(L"Could not write to device\r\n");
        }
    });
}

void MetaWearBoard::readGattCharacteristic(const void* caller, const MblMwGattChar* characteristic) {
    Guid service = byteArrayToGuid((const uint8_t*)characteristic), guidChar = byteArrayToGuid((const uint8_t*)(&characteristic->uuid_high));
    auto src = cppToWrapper.at(caller)->device->GetGattService(service)->GetCharacteristics(guidChar)->First()->Current;

    concurrency::create_task(src->ReadValueAsync()).then([&caller, &characteristic](GattCommunicationStatus status, GattReadResult result) {
        if (status != GattCommunicationStatus::Success) {
            OutputDebugString(L"Could not read gatt characteristic\r\n");
        } else {
            Array<byte>^ value;
            CryptographicBuffer::CopyToByteArray(result.Value, &value);
            mbl_mw_metawearboard_char_read(cppToWrapper.at(caller)->cppBoard, characteristic, value->Data, value->Length);
        }
    });
}

MetaWearBoard^ MetaWearBoard::getInstance(Windows::Devices::Bluetooth::BluetoothLEDevice ^device) {
    if (!instances.count(device->BluetoothAddress)) {
        instances.emplace(device->BluetoothAddress, ref new MetaWearBoard(device));
    }
    return instances.at(device->BluetoothAddress);
}

MetaWearBoard::MetaWearBoard(Windows::Devices::Bluetooth::BluetoothLEDevice ^device) {
    this->device = device;
    this->cppBoard = mbl_mw_metawearboard_create(&conn);
}

MetaWearBoard::~MetaWearBoard() {
    this->device = nullptr;
    mbl_mw_metawearboard_free(this->cppBoard);
    this->cppBoard = nullptr;
}

void MetaWearBoard::initialize() {

}
void MetaWearBoard::free() {

}